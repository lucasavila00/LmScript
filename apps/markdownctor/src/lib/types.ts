import { ParsedBlock } from "./parse-markdown";

export type GenerationInput = {
  md: string;
  parsedMd: ParsedBlock[];
  prompt: string;
};
