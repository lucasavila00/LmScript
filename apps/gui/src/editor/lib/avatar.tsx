import { ReactNode } from "react";
import { BotIcon, SparkleIcon, UserIcon } from "lucide-react";

const AVATAR_CN = "aspect-square h-full w-full m-2";
export const avatarLabel = (author: string): ReactNode => {
  switch (author) {
    case "system":
      return <SparkleIcon className={AVATAR_CN} />;
    case "user":
      return <UserIcon className={AVATAR_CN} />;
    case "assistant":
      return <BotIcon className={AVATAR_CN} />;
    default:
      return "Unk";
  }
};
export const avatarFullLabel = (author: string) => {
  switch (author) {
    case "system":
      return "System";
    case "user":
      return "User";
    case "assistant":
      return "Assistant";
    default:
      return "Unknown";
  }
};
