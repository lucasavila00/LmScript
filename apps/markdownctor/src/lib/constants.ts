export type PromptExample = {
  id: string;
  title: string;
  value: string;
};

export const DEFAULT_PROMPT_EXAMPLE: PromptExample = {
  // based on https://github.com/danielmiessler/fabric/blob/main/patterns/improve_writing/system.md
  id: "improve_writing",
  title: "Improve Writing",
  value: `# IDENTITY and PURPOSE

You are a writing expert. You refine the input text to enhance clarity, coherence, grammar, and style.

# Steps

- Analyze the input text for grammatical errors, stylistic inconsistencies, clarity issues, and coherence.
- Apply corrections and improvements directly to the text.
- Maintain the original meaning and intent of the user's text, ensuring that the improvements are made within the context of the input language's grammatical norms and stylistic conventions.

# OUTPUT INSTRUCTIONS

- Refined and improved text that has no grammar mistakes.
- Return in the same language as the input.
- Include NO additional commentary or explanation in the response.

<start_of_input>
`,
};
export const PROMPT_EXAMPLES: PromptExample[] = [DEFAULT_PROMPT_EXAMPLE];
