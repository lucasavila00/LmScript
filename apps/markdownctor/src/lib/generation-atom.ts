import { atomFamily } from "recoil";
import { GenerationInput } from "./types";
export type UiGenerationData = {
  state: "loading" | "initialized" | "finished" | "error";
  captures: Record<string, unknown>;
  finalText: string | undefined;
  error?: unknown;
};
import { AbstractBackend, Task } from "@lmscript/client/backends/abstract";
import { SGLangBackend } from "@lmscript/client/backends/sglang";
import { assertIsNever } from "./utils";
import { BULLET_LIST_REGEX, NUMBERED_LIST_REGEX } from "@lmscript/client/regex";
const getInstance = (): AbstractBackend => {
  return new SGLangBackend({
    url: "http://localhost:8080/http://localhost:30000",
    template: "mistral",
  });
};
const getTasks = (input: GenerationInput) => {
  const tasks: Task[] = [
    {
      tag: "StartRoleTask",
      role: "user",
    },
    {
      tag: "AddTextTask",
      text: input.prompt,
    },
    {
      tag: "AddTextTask",
      text: input.md,
    },
    {
      tag: "StartRoleTask",
      role: "assistant",
    },
    ...input.parsedMd.flatMap((block): Task[] => {
      switch (block.tag) {
        case "error": {
          return [
            {
              tag: "AddTextTask",
              text: block.original,
            },
          ];
        }
        case "heading": {
          return [
            {
              tag: "AddTextTask",
              text: "#".repeat(block.level),
            },
            {
              tag: "GenerateTask",
              max_tokens: 1024,
              name: block.uuid,
              stop: ["\n"],
              regex: undefined,
            },
            {
              tag: "AddTextTask",
              text: "\n\n",
            },
          ];
        }
        case "paragraph": {
          return [
            {
              tag: "GenerateTask",
              max_tokens: 1024,
              name: block.uuid,
              stop: ["\n\n"],
              regex: undefined,
            },
          ];
        }
        case "list": {
          const regex = block.ordered ? NUMBERED_LIST_REGEX : BULLET_LIST_REGEX;
          return [
            {
              tag: "GenerateTask",
              max_tokens: 1024,
              name: block.uuid,
              stop: ["\n\n"],
              regex,
            },
          ];
        }
        default: {
          return assertIsNever(block);
        }
      }
    }),
  ];

  return tasks;
};
export const generateAsyncAtom = atomFamily<
  UiGenerationData,
  {
    input: GenerationInput;
    cacheBuster: number;
  }
>({
  key: "generateAsyncAtom",
  default: (_param) => {
    return {
      state: "initialized",
      captures: {},
      finalText: undefined,
    };
  },
  effects: (param) => [
    (opts) => {
      const instance = getInstance();
      const tasks = getTasks(param.input);

      instance
        .executeJSON(
          {
            tasks,
            sampling_params: {
              temperature: 0.1,
            },
            initial_state: {
              text: "",
              captured: {},
            },
          },
          {
            onCapture: (cap) => {
              opts.setSelf((prev) => {
                if ("captures" in prev) {
                  return {
                    captures: {
                      ...prev.captures,
                      [cap.name]: String(cap.value),
                    },
                    finalText: undefined,
                    state: "loading",
                  };
                }
                return prev;
              });
            },
          },
        )
        .then((out) => {
          opts.setSelf((prev) => {
            if ("captures" in prev) {
              return {
                captures: {
                  ...prev.captures,
                  ...Object.fromEntries(
                    Object.entries(out.captured).map(([k, v]) => [k, String(v)]),
                  ),
                },
                finalText: out.text,
                state: "finished",
              };
            }
            return prev;
          });
        })
        .catch((error) => {
          opts.setSelf((prev) => {
            if ("captures" in prev) {
              return {
                ...prev,
                state: "error",
                error,
              };
            }
            return prev;
          });
        });
    },
  ],
});
