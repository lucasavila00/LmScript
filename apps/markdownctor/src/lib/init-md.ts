export const INITIAL_MARKDOWN = `### Adding text to context

Use \`.push\` to add text to the context.

Use \`.system\`, \`.user\` and \`.assistant\` to create the messages with the required
formatting for them to be assigned to their roles.

The role functions can receive a single string that will be applied to \`.push\`,
or it receives a callback that passes the client object where you can call any
of the supported functions.

\`\`\`ts
const { captured } = await client
  .system("You are a helpful assistant.")
  .user(question1)
  .assistant((m) => m.gen("answer1", { maxTokens: 256 }))
  .run();
\`\`\`

### Generation

Generates the text and captures it with a name.

\`\`\`ts
const {
  captured: { language },
} = await client.push("The best programming language is ").gen("language").run();

console.log(language);
\`\`\`

### Selection

Selects one of the choices.

\`\`\`ts
const {
  captured: { language },
} = await client
  .push("The best programming language is ")
  .select("language", { choices: ["javascript", "typescript"] })
  .run();

console.log(language);
\`\`\``;
