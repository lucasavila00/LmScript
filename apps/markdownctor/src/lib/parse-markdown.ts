import { lexer, Tokens } from "marked";

export type ParsedBlockError = {
  uuid: string;
  tag: "error";
  error: string;
  original: string;
};

export type ParsedBlockHeading = { uuid: string; tag: "heading"; level: number; content: string };

export type ParsedBlockParagraph = { uuid: string; tag: "paragraph"; content: string };

export type ParsedBlockList = { uuid: string; tag: "list"; items: string[]; ordered: boolean };

export type ParsedBlock =
  | ParsedBlockError
  | ParsedBlockHeading
  | ParsedBlockParagraph
  | ParsedBlockList;

export const parseMarkdownUnbound = (
  block: string,
  deps: {
    randomUUID: () => string;
  },
): ParsedBlock[] => {
  const parsed = lexer(block);

  return parsed.map((item) => {
    switch (item.type) {
      case "paragraph": {
        return {
          uuid: deps.randomUUID(),
          tag: "paragraph",
          content: item.text,
        };
      }
      case "list": {
        return {
          uuid: deps.randomUUID(),
          tag: "list",
          items: (item as Tokens.List).items.map((i) => i.text),
          ordered: (item as Tokens.List).ordered,
        };
      }
      case "heading": {
        return { uuid: deps.randomUUID(), tag: "heading", level: item.depth, content: item.text };
      }
      default: {
        return {
          uuid: deps.randomUUID(),
          tag: "error",
          error: String(`Type not supported: ${item.type}`),
          original: item.raw,
        };
      }
    }
  });
};

export const parseMarkdown = (block: string): ParsedBlock[] => {
  return parseMarkdownUnbound(block, { randomUUID: () => crypto.randomUUID() });
};
