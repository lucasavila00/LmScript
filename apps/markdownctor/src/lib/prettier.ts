import prettier from "prettier/standalone";
import markdown from "prettier/plugins/markdown";

export const formatNoLineBreaks = (code: string): Promise<string> =>
  prettier.format(code, { parser: "markdown", plugins: [markdown], proseWrap: "never" });
