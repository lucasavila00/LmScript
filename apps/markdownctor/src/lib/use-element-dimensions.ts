import { useCallback, useEffect, useLayoutEffect, useRef, useState } from "react";

const isServer = typeof window === "undefined";

const isSupported = !isServer && "ResizeObserver" in window;

const noop = () => {};

const noopObserver = { observe: noop, unobserve: noop };

const resizeObserver = !isSupported
  ? noopObserver
  : new ResizeObserver((entries) => {
      for (const entry of entries) {
        const { target } = entry;
        const boundingClient = target.getBoundingClientRect();
        const set = (target as $Element).$$useElementDimensionsSet;
        if (set) {
          set(Object.assign(boundingClient, entry));
        }
      }
    });

export type ElementDimensions = ResizeObserverEntry & DOMRect;

type $Element = Element & {
  $$useElementDimensionsSet?: React.Dispatch<React.SetStateAction<ElementDimensions>>;
};

const useIsomorphicLayoutEffect = isServer ? useEffect : useLayoutEffect;

// NOTE(danielkov): this is used to stub DOMRectReadonly on the server
class Rect {
  readonly bottom: number;
  readonly height: number;
  readonly left: number;
  readonly right: number;
  readonly top: number;
  readonly width: number;
  readonly x: number;
  readonly y: number;

  constructor() {
    this.bottom = 0;
    this.height = 0;
    this.left = 0;
    this.right = 0;
    this.top = 0;
    this.width = 0;
    this.x = 0;
    this.y = 0;
  }
  toJSON() {
    return JSON.stringify(this);
  }
}

const contentRect = new Rect();
const domRect = new Rect();
const size = { inlineSize: 0, blockSize: 0 };
const defaultValue: ElementDimensions = Object.assign(domRect, {
  contentBoxSize: size,
  borderBoxSize: size,
  contentRect,
  target: null as unknown as Element,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}) as any;

const useDimensions = (): [ElementDimensions, (element?: Element | null) => void] => {
  const ref = useRef<$Element>(null);

  const [dimensions, set] = useState<ElementDimensions>(defaultValue);

  const setRef = useCallback((element?: Element | null) => {
    if (ref.current) {
      resizeObserver.unobserve(ref.current);
    }
    if (element instanceof Element) {
      (element as $Element).$$useElementDimensionsSet = set;
      resizeObserver.observe(element);
    }
  }, []);

  useIsomorphicLayoutEffect(
    () => () => {
      if (ref.current) {
        resizeObserver.unobserve(ref.current);
      }
    },
    [],
  );

  return [dimensions, setRef];
};

export default useDimensions;
