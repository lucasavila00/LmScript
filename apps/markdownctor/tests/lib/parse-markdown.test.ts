import { expect, test } from "vitest";
import { parseMarkdownUnbound } from "../../src/lib/parse-markdown";
import dedent from "dedent";

const parseMarkdown = (text: string) =>
  parseMarkdownUnbound(text, {
    randomUUID: () => "uuid",
  });
test("parses heading", () => {
  const out = parseMarkdown(dedent`
    # Hello
    
    ## World

    ### third

    #### fourth

    ##### fifth

    ###### sixth
`);
  expect(out).toMatchInlineSnapshot(`
    [
      {
        "content": "Hello",
        "level": 1,
        "tag": "heading",
        "uuid": "uuid",
      },
      {
        "content": "World",
        "level": 2,
        "tag": "heading",
        "uuid": "uuid",
      },
      {
        "content": "third",
        "level": 3,
        "tag": "heading",
        "uuid": "uuid",
      },
      {
        "content": "fourth",
        "level": 4,
        "tag": "heading",
        "uuid": "uuid",
      },
      {
        "content": "fifth",
        "level": 5,
        "tag": "heading",
        "uuid": "uuid",
      },
      {
        "content": "sixth",
        "level": 6,
        "tag": "heading",
        "uuid": "uuid",
      },
    ]
  `);
});

test("parses heading and paragraph with a single line break", () => {
  const out = parseMarkdown(dedent`
    # Hello
    world
`);
  expect(out).toMatchInlineSnapshot(`
    [
      {
        "content": "Hello",
        "level": 1,
        "tag": "heading",
        "uuid": "uuid",
      },
      {
        "content": "world",
        "tag": "paragraph",
        "uuid": "uuid",
      },
    ]
  `);
});

test("parses unordered list", () => {
  const out = parseMarkdown(dedent`
    - Hello
    - world
`);
  expect(out).toMatchInlineSnapshot(`
    [
      {
        "items": [
          "Hello",
          "world",
        ],
        "ordered": false,
        "tag": "list",
        "uuid": "uuid",
      },
    ]
  `);
});

test("parses ordered list", () => {
  const out = parseMarkdown(dedent`
    1. Hello
    2. world
`);
  expect(out).toMatchInlineSnapshot(`
    [
      {
        "items": [
          "Hello",
          "world",
        ],
        "ordered": true,
        "tag": "list",
        "uuid": "uuid",
      },
    ]
  `);
});
const SAMPLE1 = `### Adding text to context

Use \`.push\` to add text to the context.

Use \`.system\`, \`.user\` and \`.assistant\` to create the messages with the required
formatting for them to be assigned to their roles.

The role functions can receive a single string that will be applied to \`.push\`,
or it receives a callback that passes the client object where you can call any
of the supported functions.

\`\`\`ts
const { captured } = await client
  .system("You are a helpful assistant.")
  .user(question1)
  .assistant((m) => m.gen("answer1", { maxTokens: 256 }))
  .run();
\`\`\`

### Generation

Generates the text and captures it with a name.

\`\`\`ts
const {
  captured: { language },
} = await client.push("The best programming language is ").gen("language").run();

console.log(language);
\`\`\`

### Selection

Selects one of the choices.

\`\`\`ts
const {
  captured: { language },
} = await client
  .push("The best programming language is ")
  .select("language", { choices: ["javascript", "typescript"] })
  .run();

console.log(language);
\`\`\``;
test("parses ordered list", () => {
  const out = parseMarkdown(SAMPLE1);
  expect(out).toMatchSnapshot();
});
