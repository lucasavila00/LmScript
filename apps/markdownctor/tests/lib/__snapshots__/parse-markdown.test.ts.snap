// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parses ordered list 2`] = `
[
  {
    "content": "Adding text to context",
    "level": 3,
    "tag": "heading",
    "uuid": "uuid",
  },
  {
    "content": "Use \`.push\` to add text to the context.",
    "tag": "paragraph",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "content": "Use \`.system\`, \`.user\` and \`.assistant\` to create the messages with the required
formatting for them to be assigned to their roles.",
    "tag": "paragraph",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "content": "The role functions can receive a single string that will be applied to \`.push\`,
or it receives a callback that passes the client object where you can call any
of the supported functions.",
    "tag": "paragraph",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: code",
    "original": "\`\`\`ts
const { captured } = await client
  .system("You are a helpful assistant.")
  .user(question1)
  .assistant((m) => m.gen("answer1", { maxTokens: 256 }))
  .run();
\`\`\`",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "content": "Generation",
    "level": 3,
    "tag": "heading",
    "uuid": "uuid",
  },
  {
    "content": "Generates the text and captures it with a name.",
    "tag": "paragraph",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: code",
    "original": "\`\`\`ts
const {
  captured: { language },
} = await client.push("The best programming language is ").gen("language").run();

console.log(language);
\`\`\`",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "content": "Selection",
    "level": 3,
    "tag": "heading",
    "uuid": "uuid",
  },
  {
    "content": "Selects one of the choices.",
    "tag": "paragraph",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: space",
    "original": "

",
    "tag": "error",
    "uuid": "uuid",
  },
  {
    "error": "Type not supported: code",
    "original": "\`\`\`ts
const {
  captured: { language },
} = await client
  .push("The best programming language is ")
  .select("language", { choices: ["javascript", "typescript"] })
  .run();

console.log(language);
\`\`\`",
    "tag": "error",
    "uuid": "uuid",
  },
]
`;
